R/PipeOpFilterRows.R

```diff
@@ -0,0 +1,171 @@
+#'   filter = expression(age < median(age) & mass > 30),
+#'   na_column = "all")
+#' )
+#' po$train(list(task))
+#' po$state
+#' @family PipeOps
+#' @include PipeOpTaskPreproc.R
+#' @export
+PipeOpFilterRows = R6Class("PipeOpFilterRows",
+  inherit = PipeOpTaskPreproc,
+  public = list(
+    initialize = function(id = "filterrows", param_vals = list()) {
+      ps = ParamSet$new(params = list(
+        ParamUty$new("filter", default = NULL, tags = c("train", "predict"), custom_check = function(x) {
+          ok = test_character(x, any.missing = FALSE, len = 1L) ||
+            is.expression(x) ||
+            test_integerish(x, lower = 1, min.len = 1L) ||
+            is.null(x)
+          if (!ok) return("Must either be a character vector of length 1, an expression, or an integerish object of row ids")
+          TRUE
+        }),
+        ParamUty$new("na_column", default = NULL, tags = c("train", "predict"), custom_check = function(x) {
+          check_character(x, any.missing = FALSE, min.len = 1L, null.ok = TRUE)
+        }),
+        ParamLgl$new("invert", default = FALSE, tags = c("train", "predict")),
+        ParamLgl$new("skip_during_predict", default = FALSE, tags = "predict"))
+      )
+      ps$values = list(filter = NULL, na_column = NULL, invert = FALSE, skip_during_predict = FALSE)
+      super$initialize(id, param_set = ps, param_vals = param_vals)
+    }
+  ),
+  private = list(
+    .na_and_filter = function(task, skip, set_state) {
+      if (skip) {
+        return(task)  # early exit if skipped (if skip_during_predict)
+      }
+
+      row_ids = task$row_ids
+
+      # NA column(s) handling
+      na = self$param_set$values$na_column
+      if (!is.null(na)) {
+        assert_subset(na, choices = c("all", colnames(task$data())))
+        if (na == "all") na = colnames(task$data())
+        na_ids = which(rowSums(is.na(task$data(cols = na))) > 0L)
+        row_ids = setdiff(row_ids, na_ids)
+      } else {
+        na_ids = integer(0L)
+      }
+
+      # filtering
+      filter = self$param_set$values$filter
+      filter_ids =
+      if (is.null(filter)) {
+        row_ids
+      } else if (is.character(filter)) {
+        assert_subset(filter, choices = task$feature_names)
+        filter_column = task$data(cols = filter)[[1L]]
+        assert_logical(filter_column)
+        which(filter_column)
+      } else if(is.expression(filter)) {
+        filter_expression = eval(filter, envir = task$data())
+        assert_logical(filter_expression, len = task$nrow)
+        which(filter_expression)
+      } else {
+        filter = as.integer(filter)
+        assert_subset(filter, choices = task$row_ids)
+        filter
+      }
+
+      row_ids = if (self$param_set$values$invert) {
+        setdiff(row_ids, filter_ids)
+      } else {
+        intersect(row_ids, filter_ids)
+      }
+
+      # only set the state if required (during training)
+      if (set_state) { 
+        self$state$na_ids = na_ids
```

Comment:
why do we make a difference between na_ids and other filtered ids?

Reply:
I kinda liked this separation so I can see directly which were `NA` / `NaN` or which were actually filtered when both is done but I am fine with merging them.